File: client-dotnet-httpclient/test/Epistola.Client.Tests/Error/ProblemDetailExceptionTest.cs
using System.Collections.Generic;
using System.Net;
using Epistola.Client.Client;
using Epistola.Client.Error;
using Epistola.Client.Model;
using Xunit;

namespace Epistola.Client.Tests.Error;

public class ProblemDetailExceptionTest
{
    private static ProblemDetailException Exception(
        ProblemDetail problem,
        IReadOnlyList<ValidationError>? errors = null,
        IReadOnlyDictionary<string, List<DataModelValidationError>>? validationErrors = null,
        string body = "{}")
        => new(
            problem,
            errors ?? new List<ValidationError>(),
            validationErrors ?? new Dictionary<string, List<DataModelValidationError>>(),
            (HttpStatusCode)problem.Status,
            body,
            headers: null);

    [Fact]
    public void TypeSlugStripsTheEpistolaTypeBase()
    {
        var ex = Exception(new ProblemDetail(
            type: "https://epistola.app/errors/not-found",
            title: "Not Found",
            status: 404,
            detail: "Tenant 'acme' was not found"));

        Assert.Equal("not-found", ex.TypeSlug);
        Assert.Equal("Not Found", ex.Title);
        Assert.Equal(404, ex.ProblemStatus);
        Assert.Equal("Tenant 'acme' was not found", ex.Detail);
        Assert.Equal("https://epistola.app/errors/not-found", ex.Type);
    }

    [Fact]
    public void TypeSlugIsNullForAboutBlankAndNonEpistolaTypes()
    {
        Assert.Null(Exception(new ProblemDetail(title: "Bad Request", status: 400)).TypeSlug);
        Assert.Null(Exception(new ProblemDetail(type: "https://example.com/oops", title: "X", status: 400)).TypeSlug);
    }

    [Fact]
    public void PlainProblemHasNoValidationErrors()
    {
        var ex = Exception(new ProblemDetail(title: "Conflict", status: 409));
        Assert.Empty(ex.Errors);
        Assert.False(ex.IsValidationProblem);
        Assert.Empty(ex.ValidationErrors);
        Assert.False(ex.IsDataModelValidationProblem);
    }

    [Fact]
    public void DataModelValidationProblemExposesPerExampleFailures()
    {
        var ex = Exception(
            new ProblemDetail(type: "https://epistola.app/errors/data-model-validation-error", title: "Data Model Validation Error", status: 422),
            validationErrors: new Dictionary<string, List<DataModelValidationError>>
            {
                ["Example 1"] = new() { new DataModelValidationError(path: "/name", message: "required property 'name' not found") },
            });

        Assert.True(ex.IsDataModelValidationProblem);
        Assert.False(ex.IsValidationProblem);
        Assert.Equal(KnownProblemSlugs.DATA_MODEL_VALIDATION_ERROR, ex.TypeSlug);
        Assert.Equal("/name", ex.ValidationErrors["Example 1"][0].Path);
    }

    [Fact]
    public void ValidationProblemExposesFieldErrors()
    {
        var ex = Exception(
            new ProblemDetail(type: "https://epistola.app/errors/validation-error", title: "Bad Request", status: 400),
            errors: new List<ValidationError>
            {
                new(field: "name", message: "must not be blank"),
                new(field: "slug", message: "invalid format", rejectedValue: "A B"),
            });

        Assert.True(ex.IsValidationProblem);
        Assert.Equal(2, ex.Errors.Count);
        Assert.Equal("name", ex.Errors[0].Field);
        Assert.Equal("validation-error", ex.TypeSlug);
    }

    [Fact]
    public void IsApiExceptionCarryingTheOriginalBodyAndStatus()
    {
        const string body = "{\"type\":\"about:blank\",\"title\":\"Conflict\",\"status\":409}";
        var ex = Exception(new ProblemDetail(title: "Conflict", status: 409), body: body);

        // Assignable to the generated ApiException so existing catch sites keep working.
        ApiException asParent = ex;
        Assert.Equal(body, asParent.ErrorContent);
        Assert.Equal(409, asParent.ErrorCode);
    }
}

File: contracts/api/clients/kotlin-spring-restclient/src/test/kotlin/app/epistola/client/error/ProblemDetailExceptionTest.kt
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

package app.epistola.client.error

import app.epistola.client.model.DataModelValidationError
import app.epistola.client.model.ProblemDetail
import app.epistola.client.model.ValidationError
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.web.client.RestClientResponseException
import java.net.URI
import java.nio.charset.StandardCharsets
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

class ProblemDetailExceptionTest {

    private fun exception(
        problem: ProblemDetail,
        errors: List<ValidationError> = emptyList(),
        validationErrors: Map<String, List<DataModelValidationError>> = emptyMap(),
        body: String = "{}",
    ) = ProblemDetailException(
        problem = problem,
        errors = errors,
        validationErrors = validationErrors,
        statusCode = HttpStatus.valueOf(problem.status),
        statusText = HttpStatus.valueOf(problem.status).reasonPhrase,
        headers = HttpHeaders(),
        responseBody = body.toByteArray(),
        responseCharset = StandardCharsets.UTF_8,
    )

    @Test
    fun `typeSlug strips the Epistola type base`() {
        val ex = exception(
            ProblemDetail(
                type = URI.create("https://epistola.app/errors/not-found"),
                title = "Not Found",
                status = 404,
                detail = "Tenant 'acme' was not found",
            ),
        )
        assertEquals("not-found", ex.typeSlug)
        assertEquals("Not Found", ex.title)
        assertEquals(404, ex.problemStatus)
        assertEquals("Tenant 'acme' was not found", ex.detail)
        assertEquals(URI.create("https://epistola.app/errors/not-found"), ex.type)
    }

    @Test
    fun `typeSlug is null for about blank and non-Epistola types`() {
        assertNull(exception(ProblemDetail(title = "Bad Request", status = 400)).typeSlug)
        assertNull(
            exception(
                ProblemDetail(type = URI.create("https://example.com/oops"), title = "X", status = 400),
            ).typeSlug,
        )
    }

    @Test
    fun `plain problem has no validation errors`() {
        val ex = exception(ProblemDetail(title = "Conflict", status = 409))
        assertTrue(ex.errors.isEmpty())
        assertFalse(ex.isValidationProblem)
        assertTrue(ex.validationErrors.isEmpty())
        assertFalse(ex.isDataModelValidationProblem)
    }

    @Test
    fun `data-model validation problem exposes per-example failures`() {
        val ex = exception(
            ProblemDetail(
                type = URI.create("https://epistola.app/errors/data-model-validation-error"),
                title = "Data Model Validation Error",
                status = 422,
            ),
            validationErrors = mapOf(
                "Example 1" to listOf(DataModelValidationError(path = "/name", message = "required property 'name' not found")),
            ),
        )
        assertTrue(ex.isDataModelValidationProblem)
        assertFalse(ex.isValidationProblem)
        assertEquals(KnownProblemSlugs.DATA_MODEL_VALIDATION_ERROR, ex.typeSlug)
        assertEquals("/name", ex.validationErrors.getValue("Example 1")[0].path)
    }

    @Test
    fun `validation problem exposes field errors`() {
        val ex = exception(
            ProblemDetail(
                type = URI.create("https://epistola.app/errors/validation-error"),
                title = "Bad Request",
                status = 400,
            ),
            errors = listOf(
                ValidationError(field = "name", message = "must not be blank"),
                ValidationError(field = "slug", message = "invalid format", rejectedValue = "A B"),
            ),
        )
        assertTrue(ex.isValidationProblem)
        assertEquals(2, ex.errors.size)
        assertEquals("name", ex.errors[0].field)
        assertEquals("validation-error", ex.typeSlug)
    }

    @Test
    fun `is a RestClientResponseException carrying the original body`() {
        val body = """{"type":"about:blank","title":"Conflict","status":409}"""
        val ex = exception(ProblemDetail(title = "Conflict", status = 409), body = body)
        // Assignable to the generated @Throws type so existing catch sites keep working.
        val asParent: RestClientResponseException = ex
        assertEquals(body, asParent.responseBodyAsString)
        assertEquals(HttpStatus.CONFLICT, asParent.statusCode)
    }
}
